Update SDK v2 migration guide.md (#2889)
Azure/azure-iot-sdk-csharp: SDK v2 migration guide.md

  - Removing or upgrading several NuGet dependencies.
    - Upgraded
      - Microsoft.Azure.Devices.Client
      - Microsoft.Azure.Devices
      - Microsoft.Azure.Devices.Authentication
      - Microsoft.Azure.Devices.Provisioning.Client
      - Microsoft.Azure.Devices.Provisioning.Service
    - Removed
      - Microsoft.Azure.Devices.Shared
      - Microsoft.Azure.Devices.Provisioning.Transport.Amqp
      - Microsoft.Azure.Devices.Provisioning.Transport.Http
      - Microsoft.Azure.Devices.Provisioning.Transport.Mqtt
      - Microsoft.Azure.Devices.Provisioning.Security.Tpm
  - Consolidating IoT hub service clients and renaming to reflect the items or operations they support.
    - Many existing client classes (RegistryManager, ServiceClient, etc.) were confusingly named and contained methods that weren't always consistent with the client's assumed responsibilities.
  - Many existing clients had a mix of standard constructors (`new DeviceClient(...)`) and static builder methods (`DeviceClient.CreateFromConnectionString(...)`) that caused some confusion among users. The factory methods have been removed and the addition of constructors in clients enables unit testing.

>   - Exception handling. (Exception handling in the v1 SDK required multiple exception types to be caught. This logic has been consolidated to always throw only `IotHubClientException` for exceptions arising from communication attempts with IoT Hub. Parameter validation will throw `ArgumentException`/`ArgumentNullException`/`FormatException`/`InvalidOperationException`.

Reviewer: ```suggestion
  - Exception handling.
     - Exception handling in the v1 SDK required multiple exception types to be caught. This logic has been consolidated to always throw `IotHubClientException` and `IotHubServiceException` for exceptions arising from communication attempts with IoT Hub, `DeviceProvisioningClientException` and `DeviceProvisioningServiceException` for exceptions arising from communication attempts with DPS. Parameter validation will throw `ArgumentException`/`ArgumentNullException`/`FormatException`/`InvalidOperationException`.
```

Reviewer: Also, currently we still have several other types of exception thrown by our clients, see [here](https://github.com/Azure/azure-iot-sdk-csharp/pull/2887#issuecomment-1281563556).  Do we want to consolidate them as well? @drwill-ms @abhipsaMisra @timtay-microsoft

Reviewer 2: Document them here? Yes.

Reviewer: > Document them here?

No, more for the question whether we want to map those exceptions into the hub/dps exceptions or keep throwing them as-is?

Reviewer 3: I think we should keep our exception list as narrow as possible. SwaggerException, HttpRequestException and CryptographicException should all fit into IotHub(Client/Service)Exception, since they arise out of communication attempts with IoT hub. As for the IndexOutOfRangeException, how deep in the stack do we throw this? Would it fit into our "parameter validation" exception?

Reviewer: > As for the IndexOutOfRangeException, how deep in the stack do we throw this? Would it fit into our "parameter validation" exception?

The only 2 [places ](https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/provisioning/service/src/Models/Query.cs#L135
)of throwing `IndexOutOfRangeException` are both in the model "Query.cs". It may be fit into `InvalidOperationException` according to the use case here. Thoughts?

Reviewer 2: For Query.cs, line 135 why not ArgumentNullException (null) and ArgumentException (whitespace)?

Reviewer 2: For Query.cs line 153, I'd agree with InvalidOperationException.

Reviewer: Yes, I agree.

   - [IoT hub service client](#iot-hub-service-client)
   - [Device Provisioning Service (DPS) device client](#dps-device-client)
   - [DPS service client](#dps-service-client)
   - [Security provider client](#security-provider-client)
 - [Frequently asked questions](#frequently-asked-questions)

## Why the version 1 SDK is being replaced

There are a number of reasons why the Azure IoT SDK team chose to do a major version revision. Here are a few of the more important reasons:
  - Removing or upgrading several NuGet dependencies.
    - Upgraded
      - Microsoft.Azure.Devices.Client
      - Microsoft.Azure.Devices

>       - Microsoft.Azure.Devices.Authentication

Reviewer 2: Auth is a new one, extracted from a previous package.

      - Microsoft.Azure.Devices.Provisioning.Client
      - Microsoft.Azure.Devices.Provisioning.Service
    - Removed
      - Microsoft.Azure.Devices.Shared
      - Microsoft.Azure.Devices.Provisioning.Transport.Amqp
      - Microsoft.Azure.Devices.Provisioning.Transport.Http
      - Microsoft.Azure.Devices.Provisioning.Transport.Mqtt
      - Microsoft.Azure.Devices.Provisioning.Security.Tpm
  - Consolidating IoT hub service clients and renaming to reflect the items or operations they support.
    - Many existing client classes (RegistryManager, ServiceClient, etc.) were confusingly named and contained methods that weren't always consistent with the client's assumed responsibilities.
  - Many existing clients had a mix of standard constructors (`new DeviceClient(...)`) and static builder methods (`DeviceClient.CreateFromConnectionString(...)`) that caused some confusion among users. The factory methods have been removed and the addition of constructors in clients enables unit testing.
  - Exception handling.

>     - Exception handling in the v1 SDK required multiple exception types to be caught. This logic has been consolidated to always throw `IotHubClientException` and `IotHubServiceException` for exceptions arising from communication attempts with IoT Hub, `DeviceProvisioningClientException` and `DeviceProvisioningServiceException` for exceptions arising from communication attempts with DPS. Parameter validation will throw `ArgumentException`/`ArgumentNullException`/`FormatException`/`InvalidOperationException`.

Reviewer 2: Worth clarifying that one is for service client and one is for device client. Neither will throw the other's exception.

>     - Exception handling in the v1 SDK required multiple exception types to be caught. This logic has been consolidated to always throw `IotHubClientException` and `IotHubServiceException` for exceptions arising from communication attempts with IoT Hub, `DeviceProvisioningClientException` and `DeviceProvisioningServiceException` for exceptions arising from communication attempts with DPS. Parameter validation will throw `ArgumentException`/`ArgumentNullException`/`FormatException`/`InvalidOperationException`.

Reviewer 2: Also, OperationCanceledException. Another recent find is ArgumentOutOfRangeException.

Worth mentioning that ArgumentNullException and ArgumentOutOfRangeException both inherit from ArgumentException. That way you mention ArgumentNullException and then in parenthesis say, "and also these that inherit from it and provide more specific feedback".

   - [IoT hub service client](#iot-hub-service-client)
   - [Device Provisioning Service (DPS) device client](#dps-device-client)
   - [DPS service client](#dps-service-client)
   - [Security provider client](#security-provider-client)
 - [Frequently asked questions](#frequently-asked-questions)

## Why the version 1 SDK is being replaced

There are a number of reasons why the Azure IoT SDK team chose to do a major version revision. Here are a few of the more important reasons:

> ### Creating, removing, or upgrading several NuGet dependencies.

Reviewer 2: ```suggestion
### Creating, removing, or upgrading several NuGet packages
```

  - Created
    - Microsoft.Azure.Devices.Authentication
  - Upgraded
    - Microsoft.Azure.Devices (IoT hub service)
    - Microsoft.Azure.Devices.Client (IoT hub device)
    - Microsoft.Azure.Devices.Provisioning.Client
    - Microsoft.Azure.Devices.Provisioning.Service
  - Removed
    - Microsoft.Azure.Devices.Shared
    - Microsoft.Azure.Devices.Provisioning.Transport.Amqp
    - Microsoft.Azure.Devices.Provisioning.Transport.Http
    - Microsoft.Azure.Devices.Provisioning.Transport.Mqtt
    - Microsoft.Azure.Devices.Provisioning.Security.Tpm

### Consolidating IoT hub service clients and renaming to reflect the items or operations they support.

  Many existing client classes (RegistryManager, ServiceClient, etc.) were confusingly named and contained methods that weren't always consistent with the client's assumed responsibilities.

### Client constructors and mocking

>   Many existing clients had a mix of standard constructors (`new DeviceClient(...)`) and static builder methods (`DeviceClient.CreateFromConnectionString(...)`) that caused some confusion among users. The factory methods have been removed and the addition of constructors in clients also enables unit test mocking.

Reviewer 2: ```suggestion
Many existing clients had a mix of standard constructors (`new RegistryManager(...)`) and static builder methods (`DeviceClient.CreateFromConnectionString(...)`) that caused some confusion among users. The factory methods have been removed and the addition of constructors in clients also enables unit test mocking.
```

### Exception handling

>   - Exception handling in the v1 SDK required a wide variety of exception types to be caught. Exception types have been consolidated.
>     - Parameter validation may throw the following exceptions.
>       - `ArgumentException` for basic parameter validation.
>         - Also, `ArgumentNullException` and `ArgumentOutOfRangeException` inherit from it and provide more specific feedback.
>       - `FormatException` when a string parameter format does not match what is expected (e.g., connection string with embedded key/value pairs).
>       - `InvalidOperationException`
>         - For example, when calling device client operations before explicility calling `IotHubDeviceClient.OpenAsync()` first.
>         - When some runtime validated user input is invalid in some other way, for example, importing or exporting devices where there is a null value in the list.
>     - `OperationCanceledException` when a cancellation token is signaled.
>     - When an operation fails:
>       - `IotHubClientException` for device client and `IotHubServiceException` for service client for any exceptions arising from communication attempts with IoT hub.
>         - Based on `IotHubServiceErrorCode`, we determine if an exception is transient. Check error code for a specific error in details.
>       - `DeviceProvisioningClientException` for provisioning client and `DeviceProvisioningServiceException` for provisioning service client for exceptions arising from communication attempts with DPS.

Reviewer 2: ```suggestion
Exception handling in the v1 SDK required a wide variety of exception types to be caught. Exception types have been consolidated.

- Parameter validation may throw the following exceptions.
  - `ArgumentException` for basic parameter validation.
    - Also, `ArgumentNullException` and `ArgumentOutOfRangeException` inherit from it and provide more specific feedback.
  - `FormatException` when a string parameter format does not match what is expected (e.g., connection string with embedded key/value pairs).
  - `InvalidOperationException`
    - For example, when calling device client operations before explicility calling `IotHubDeviceClient.OpenAsync()` first.
    - When some runtime validated user input is invalid in some other way, for example, importing or exporting devices where there is a null value in the list.
- When cancellation is requested.
  - `OperationCanceledException` when a user-supplied cancellation token is signaled.
- When an operation fails:
  - `IotHubClientException` for device client and `IotHubServiceException` for service client for any exceptions arising from communication attempts with IoT hub.
    - Based on `IotHubServiceErrorCode`, we determine if an exception is transient. Check error code for a specific error in details.
  - `DeviceProvisioningClientException` for provisioning client and `DeviceProvisioningServiceException` for provisioning service client for exceptions arising from communication attempts with DPS.
```

> ### Connection monitoring and client lifetime
>   - Caching the latest connection status information in device client.
>   - `RecommendedAction` in connection status handling based on connection status.

Reviewer 2: ```suggestion
### Connection monitoring and client lifetime

- Caching the latest connection status information on device client, so the client app does not have to do it.
- `RecommendedAction` enum provided in connection status handling based on connection status to assist in know what action the client app is recommended to take.
```

| Version 1 API | Equivalent version 2 API |
|:---|:---|

> | `RegistryManager` | `IotHubServiceClient`, `subclients`, `Devices`, `Twins`, `Configurations`, etc. |

Reviewer 2: ```suggestion
| `RegistryManager` | `IotHubServiceClient`, subclients: `Devices`, `Twins`, `Configurations`, etc. |
```

Reviewer 2: subclients isn't an actual type (meaning we don't need the code (``) wrapper around it.
